### PR TITLE
Update live kernel models on remove document open

### PR DESCRIPTION
**Pull request by IanMatthewHuff**
For #5984 

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

**Review by DavidKutu — APPROVED** [review]

**Review by DonJayamanne — APPROVED** [review]

File: src/client/datascience/notebook/notebookControllerManager.ts [hunk 1]
```diff
@@ -231,6 +257,9 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
         document: NotebookDocument,
         token: CancellationToken
     ): Promise<KernelConnectionMetadata | undefined> {
+        // Don't look for a preferred kernel until we have finished our initial controller load
```
Comment (rchiodo) [inline src/client/datascience/notebook/notebookControllerManager.ts:260]: This seems out of place here. It has nothing to do with finding a preferred kernel. Perhaps the caller of this should make this call?

Comment (IanMatthewHuff) [reply src/client/datascience/notebook/notebookControllerManager.ts:260]: @rchiodo Sorry I had just been waiting for Don's review to push, so I pushed before this comment came in. I actually agree with you, it's a precondition for setting the preferredController, but it's not a precondition for finding a preferred kernel. I'll do a follow up PR.

File: src/client/datascience/notebook/notebookControllerManager.ts [hunk 2]
```diff
@@ -173,35 +183,51 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
...
+        } else {
+            this.findPreferredController(document, preferredSearchToken.token).finally(() => {
                 // Make sure that we clear our finding in progress when done
                 this.findPreferredInProgress.delete(document);
             });
+        }
+    }
+
+    // Find the preferred controller for the given notebook document
+    private async findPreferredController(document: NotebookDocument, cancelToken: CancellationToken) {
```
Comment (rchiodo) [inline src/client/datascience/notebook/notebookControllerManager.ts:211]: Name feels off to me. It's setting the controller mapping, right? That's the goal of this function? Maybe setController?

File: src/client/datascience/notebook/notebookControllerManager.ts [hunk 3]
```diff
@@ -60,6 +60,9 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
     // Promise to resolve when we have loaded our controllers
     private controllersPromise?: Promise<VSCodeNotebookController[]>;
 
+    // Listing of the controllers that we have registered
+    private registeredControllers: VSCodeNotebookController[] = [];
```
Comment (rchiodo) [inline src/client/datascience/notebook/notebookControllerManager.ts:64]: I don't think this is necessary. You only use it one function. And in that function you have to wait on the controllers promise anyway. You could just use the result of getNotebookControllers instead.

**Review by rchiodo — CHANGES_REQUESTED** [review]
⏲️

**Pull request merged**